platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/DefaultCompositeTestEventGenerator.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.tasks.testing.CompositeTestEventGenerator;
+import org.gradle.api.tasks.testing.TestEventGenerator;
+import org.gradle.internal.id.IdGenerator;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.WeakHashMap;
+
+@NonNullApi
+public class DefaultCompositeTestEventGenerator extends DefaultTestEventGenerator implements CompositeTestEventGenerator {
+    private final IdGenerator<?> idGenerator;
+    private Set<DefaultTestEventGenerator> children;
+
+    public DefaultCompositeTestEventGenerator(
+        TestResultProcessor processor, IdGenerator<?> idGenerator, @Nullable TestDescriptorInternal parent, TestDescriptorInternal testDescriptor
+    ) {
+        super(processor, parent, testDescriptor);
+        this.idGenerator = idGenerator;
+    }
+
+    private void addChild(DefaultTestEventGenerator child) {
+        if (children == null) {
+            children = Collections.newSetFromMap(new WeakHashMap<>());
+        }
```

Comment:
(1) Holdover from when I initially merged the two types.
(2) I'm not sure of the scale of test events that would be generated, but I don't want to hold extra references to child generators here. It would probably be better if I explicitly removed the entry by giving a reference to the child though.